### PR TITLE
20 geoserver v2

### LINKED ISSUE #20: Geoserver setup and configuration
UI:
https://github.com/HSLdevcom/jore-map-ui/pull/234
Backend:
https://gitlab.hsl.fi/jore/jore-map-backend/merge_requests/6

### PR DESCRIPTION
**Pull request by kronken**
Closes #20

### CODE REVIEW

File: src/stores/networkStore.ts [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+        this._showLinks = false;
+        this._showNodes = false;
+        this._showPoints = false;
+    }
+
+    private isAnythingEnabled() {
+        return this._showLinks || this._showNodes;
+    }
+
+    @computed get showLinks(): boolean {
```
Comment (Divinare) [inline src/stores/networkStore.ts]: Generally a verb is a bad name for a boolean. "isLinksVisible / isLinksShown" would be a better name

File: src/stores/networkStore.ts [hunk 2]
```diff
@@ -0,0 +1,77 @@
+import { action, computed, observable } from 'mobx';
+import TransitType from '../enums/transitType';
+
+export class NetworkStore {
+    @observable private _enabledTypes: TransitType[];
```
Comment (Divinare) [inline src/stores/networkStore.ts]: Somewhere in our code we use verb "select" and somewhere "enable". For example we call this variable "selectedTypes" and somewhere "enabledTypes". Pick 1 as a convention. I would personally use "select, selected etc."

File: src/stores/networkStore.ts [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+            TransitType.SUBWAY,
+            TransitType.TRAIN,
+            TransitType.TRAM,
+        ];
+        this._showLinks = false;
+        this._showNodes = false;
+        this._showPoints = false;
+    }
+
+    private isAnythingEnabled() {
```
Comment (Divinare) [inline src/stores/networkStore.ts]: isAnythingEnabled is too generic name.

File: src/util/transitTypeHelper.tsx [hunk 4]
```diff
@@ -0,0 +1,20 @@
+import TransitType from '../enums/transitType';
+
+export default class TransitTypeHelper {
+    public static convertTransitTypeCodeToTransitType = (type: string) => {
```
Comment (Divinare) [inline src/util/transitTypeHelper.tsx]: Is this helper really needed? Can we capsule this inside factories?

File: src/components/controls/TransitToggleButtonBar.tsx [hunk 5]
```diff
@@ -1,91 +1,42 @@
...
-                    toggleActivity={this.toggleActivity}
-                    toggled={this.state.toggles.subway}
+                    toggleActivity={this.props.toggleSelectedTypes.bind(this, TransitType.SUBWAY)}
+                    toggled={this.props.selectedTypes.includes(TransitType.SUBWAY)}
                     type={TransitType.SUBWAY}
                 />
                 <TransitToggleButton
-                    toggleActivity={this.toggleActivity}
-                    toggled={this.state.toggles.ferry}
+                    toggleActivity={this.props.toggleSelectedTypes.bind(this, TransitType.FERRY)}
```
[the remark this answers is not in the document]
Comment (neodyymi) [reply src/components/controls/TransitToggleButtonBar.tsx]: Seems to work just with
` toggleActivity={this.props.toggleSelectedTypes}`

[the remark this answers is not in the document]
Comment (kronken) [reply src/components/controls/TransitToggleButtonBar.tsx]: fixed

File: src/components/map/layers/NetworkLayers.tsx [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+import { observer, inject } from 'mobx-react';
+import VectorgridLayer from './VectorgridLayer';
+import { NetworkStore } from '../../../stores/networkStore';
+import TransitTypeHelper from '../../../util/transitTypeHelper';
+import TransitTypeColorHelper from '../../../util/transitTypeColorHelper';
+import NodeType from '../../../enums/nodeType';
+
+const layerNameSeparator = '{LAYER_NAME}';
+// tslint:disable-next-line:max-line-length
+const geoserverVectorTileLayerAddress = `http://localhost/geoserver/gwc/service/tms/1.0.0/joremapui%3A${layerNameSeparator}@EPSG%3A900913@pbf/{z}/{x}/{y}.pbf`;
```
[the remark this answers is not in the document]
Comment (kronken) [reply src/components/map/layers/NetworkLayers.tsx]: I'm actually building logic for this coming from environment_variables, in order to allow dynamic urls on server. This is happening in: https://github.com/HSLdevcom/jore-map-ui/issues/235

File: src/util/transitTypeHelper.tsx (hunk 4 above)
Comment (kronken) [reply src/util/transitTypeHelper.tsx]: Currently we also use it outside of the factory, so in that sense we need it. This however relates to if NetworkLayers.tsx logic should be in factory or not, which is mentioned in https://github.com/HSLdevcom/jore-map-ui/pull/234#discussion_r221507893

File: src/components/map/layers/NetworkLayers.tsx [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+
+enum geoserverLayer {
+    Node = 'solmu',
+    Link = 'linkki',
+    Point = 'piste',
+}
+
+enum nodeColors {
```
[the remark this answers is not in the document]
Comment (kronken) [reply src/components/map/layers/NetworkLayers.tsx]: Hmm, this relates to if we see the network color of points as isolated. Personally, I can understand both reasonings.

File: src/components/map/layers/NetworkLayers.tsx [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+
+interface INetworkLayersProps {
+    networkStore?: NetworkStore;
+}
+
+@inject('networkStore')
+@observer
+export default class NetworkLayers extends Component<INetworkLayersProps> {
+    private getAddress(layer: geoserverLayer) {
+        return geoserverVectorTileLayerAddress.replace(layerNameSeparator, layer);
```
[the remark this answers is not in the document]
Comment (kronken) [reply src/components/map/layers/NetworkLayers.tsx]: In every case, this is going to change to using env varibles in: https://github.com/HSLdevcom/jore-map-ui/issues/235

File: src/components/map/layers/NetworkLayers.tsx [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+                    color,
+                    weight: 1,
+                    fillOpacity: 1,
+                    fill: true,
+                };
+            },
+        };
+    }
+
+    private getPointStyle = () => {
```
[the remark this answers is not in the document]
Comment (kronken) [reply src/components/map/layers/NetworkLayers.tsx]: This is pretty much how styles are created for vectors:
https://leaflet.github.io/Leaflet.VectorGrid/vectorgrid-api-docs.html#styling-vectorgrids

File: src/components/map/layers/NetworkLayers.tsx [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+
+                return {
+                    color,
+                    radius: 1,
+                };
+            },
+        };
+    }
+
+    private getNodeStyle = () => {
```
[the remark this answers is not in the document]
Comment (kronken) [reply src/components/map/layers/NetworkLayers.tsx]: Check https://github.com/HSLdevcom/jore-map-ui/pull/234#discussion_r221549999

File: src/components/map/layers/NetworkLayers.tsx [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+                    color,
+                    radius: 2,
+                };
+            },
+        };
+    }
+
+    render() {
+        return (
+            <React.Fragment>
```
[the remark this answers is not in the document]
Comment (kronken) [reply src/components/map/layers/NetworkLayers.tsx]: This is basically the same things:
https://github.com/HSLdevcom/jore-map-ui/pull/234#discussion_r221549999

[the remark this answers is not in the document]
Comment (kronken) [reply src/components/map/layers/NetworkLayers.tsx]: As told in the docs, they recommend using React.Fragments if we depend on tools not supporting <>. But, if it has previously worked, then I can change to use <>

File: src/components/sidebar/homeView/HomeView.tsx [hunk 12]
```diff
@@ -5,24 +5,31 @@ import TransitToggleButtonBar from '../../controls/TransitToggleButtonBar';
...
 
 @inject('searchStore')
 @observer
-class HomeView extends React.Component<ISidebarProps> {
-    private setFiltersFunction = (filters: string[]): void => {
-        this.props.searchStore!.filters = filters;
+class HomeView extends React.Component<IHomeViewProps> {
+    constructor (props: IHomeViewProps) {
+        super(props);
+        this.toggleSelectedType = this.toggleSelectedType.bind(this);
```
[the remark this answers is not in the document]
Comment (kronken) [reply src/components/sidebar/homeView/HomeView.tsx]: I think it is :D

File: src/stores/searchStore.ts [hunk 13]
```diff
@@ -48,12 +55,17 @@ export class SearchStore {
...
-    @computed get filters(): string[] {
-        return this._filters;
+    @computed get selectedTypes(): TransitType[] {
+        return this._selectedTypes;
     }
 
-    set filters(value: string[]) {
-        this._filters = value;
+    @action
+    public toggleSelectedTypes(type: TransitType) {
```
[the remark this answers is not in the document]
Comment (kronken) [reply src/stores/searchStore.ts]: Hmm. I think toggleType in a searchStore context is not declarative enough.

File: src/util/transitTypeColorHelper.tsx [hunk 14]
```diff
@@ -2,6 +2,25 @@ import TransitType from '../enums/transitType';
 import * as s from './transitTypeColors.scss';
 
 class TransitTypeColorHelper {
+    public static getColor = (type: TransitType) => {
```
[the remark this answers is not in the document]
Comment (kronken) [reply src/util/transitTypeColorHelper.tsx]: Maybe, but that would probably be rather hacky. This is however how vectorgrid styles are done by the docs.

File: src/components/map/layers/NetworkLayers.tsx [hunk 15]
```diff
@@ -0,0 +1,119 @@
...
+                    color,
+                    radius: 1,
+                };
+            },
+        };
+    }
+
+    private getNodeStyle = () => {
+        return {
+            solmu: (properties: any, zoom: any) => {
```
[the remark this answers is not in the document]
Comment (Divinare) [reply src/components/map/layers/NetworkLayers.tsx]: How about: "Jore column name directly mirrored as layer in geoserver"

File: src/components/map/layers/VectorGridLayer.tsx [hunk 16]
```diff
@@ -0,0 +1,31 @@
...
+import { GridLayer, GridLayerProps, withLeaflet } from 'react-leaflet';
+
+declare module 'leaflet' {
+    let vectorGrid: any;
+}
+
+interface IVectorGridLayerProps extends GridLayerProps {
+    url: string;
+    tms: boolean;
+    vectorTileLayerStyles: any;
```
Comment (Divinare) [inline src/components/map/layers/VectorGridLayer.tsx]: Is this variable used?

File: src/components/map/layers/NetworkLayers.tsx (hunk 15 above)
[the remark this answers is not in the document]
Comment (kronken) [reply src/components/map/layers/NetworkLayers.tsx]: Setting: Layer name 'linkki' is directly mirrored from JORE through geoserver

File: src/components/sidebar/routesView/RoutesList.tsx [hunk 17]
```diff
@@ -56,8 +46,20 @@ class RoutesList extends React.Component<IRoutesListProps, IRoutesListState> {
         }
     }
 
-    private setFiltersFunction = (filters: string[]) => {
-        // TODO: Make this do something to network.
+    public toggleSelectedTransitTypes = (type: TransitType) => {
```
Comment (Divinare) [inline src/components/sidebar/routesView/RoutesList.tsx]: toggleSelectedTransitTypes -> toggleTransitType?

File: src/components/sidebar/homeView/HomeView.tsx [hunk 18]
```diff
@@ -5,24 +5,26 @@ import TransitToggleButtonBar from '../../controls/TransitToggleButtonBar';
...
     searchStore?: SearchStore;
 }
 
 @inject('searchStore')
 @observer
-class HomeView extends React.Component<ISidebarProps> {
-    private setFiltersFunction = (filters: string[]): void => {
-        this.props.searchStore!.filters = filters;
+class HomeView extends React.Component<IHomeViewProps> {
+    public toggleSelectedType = (type: TransitType) => {
```
Comment (Divinare) [inline src/components/sidebar/homeView/HomeView.tsx]: toggleTransitType

File: src/components/sidebar/routesView/RoutesView.tsx [hunk 19]
```diff
@@ -13,27 +13,28 @@ import navigator from '../../../routing/navigator';
...
     routeStore?: RouteStore;
     searchStore?: SearchStore;
     sidebarStore?: SidebarStore;
 }
 
 @inject('routeStore', 'searchStore', 'sidebarStore')
 @observer
-class RoutesView extends React.Component<ISidebarProps> {
+class RoutesView extends React.Component<IRoutesViewProps> {
+    public toggleSelectedType = (type: TransitType) => {
```
Comment (Divinare) [inline src/components/sidebar/routesView/RoutesView.tsx]: toggleTransitType

File: src/components/map/layers/VectorGridLayer.tsx (hunk 16 above)
Comment (kronken) [reply src/components/map/layers/VectorGridLayer.tsx]: Check: https://github.com/HSLdevcom/jore-map-ui/pull/234/files#diff-612cdc0c79c52d292f22d01e01e4ed78R101

File: src/stores/networkStore.ts [hunk 20]
```diff
@@ -0,0 +1,66 @@
...
+    public toggleShowLinks() {
+        this._isLinksVisible = !this._isLinksVisible;
+    }
+
+    @computed get isNodesVisible(): boolean {
+        return this._isNodesVisible;
+    }
+
+    @action
+    public toggleShowNodes() {
```
Comment (Divinare) [inline src/stores/networkStore.ts]: nit: A better name would be toggleIsNodesVisible

File: src/stores/networkStore.ts [hunk 21]
```diff
@@ -0,0 +1,66 @@
...
+import TransitType from '../enums/transitType';
+
+export class NetworkStore {
+    @observable private _selectedTypes: TransitType[];
+    @observable private _isLinksVisible: boolean;
+    @observable private _isNodesVisible: boolean;
+    @observable private _isPointsVisible: boolean;
+
+    constructor() {
+        this._selectedTypes = [
```
Comment (Divinare) [inline src/stores/networkStore.ts]: _selectedTransitTypes?

**Comment by neodyymi** [comment]
Looks good! Merging..

**Pull request merged**